fix(learn): colored markers - step 5 - validation

Saul-BT opened the pull request:
Checklist:

<!-- Please follow this checklist and put an x in each of the boxes, like this: [x]. It will ensure that our team takes your pull request seriously. -->

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

<!--If your pull request closes a GitHub issue, replace the XXXXX below with the issue number.-->

Closes #47805

<!-- Feel free to add any additional description of changes below this line -->

Saul-BT committed:
fix: optional whitespaces validation

Fixes #47805

jeremylt commented on curriculum/challenges/english/14-responsive-web-design-22/learn-css-colors-by-building-a-set-of-colored-markers/616965351e74d4689eb6de30.md:
````diff
@@ -30,7 +30,7 @@ Your new `meta` element should have a `name` attribute set to `viewport`, and a
 ```js
 const meta = [...document.querySelectorAll('meta')];
-const target = meta?.find(m => m?.getAttribute('name') === 'viewport' && m?.getAttribute('content') === 'width=device-width, initial-scale=1.0');
+const target = meta?.find(m => m?.getAttribute('name') === 'viewport' && / *width *= *device-width *, *initial-scale *= *1\.0 */.test(m?.getAttribute('content')));
````
It might be better to trim instead of introducing a regex

Saul-BT replied:
Hmmm, well, for this particular case I don't think so, besides it may add more complexity to the code.

This fix also works for other valid cases like:
`"width = device-width , initial-scale=1.0 "`

PD: But with your comment, I realized that I can improve the solution by instantiating the regex outside the `find`, so that I don't create it in each iteration. 😅

jeremylt replied:
Ah, I missed the internal white space

Saul-BT committed:
Improve: use one instance of the regex

raisedadead commented:
As a new contributor, we encourage you to read our contributing guidelines. 

We expect our contributors to be aware of the process specific to this project. Following the guidelines religiously earns you the respect of fellow maintainers and saves everyone time.

Some examples of this are:

1. Do not edit off of GitHub–while you can, it's not a good idea.
2. Make sure you follow the PR checklist and not just tick things off; otherwise, we won't take you seriously.
3. Use the correct way to link PRs in the description of the PR by updating the `XXXXXX`. Do not just add issue numbers everywhere and anywhere you feel like.
4. Use mentions in comments or review requests to a minimum or not. We understand you are excited about contributing, and our maintainers will get back to you as soon as they get a chance.

We appreciate you taking the time to help us, and we hope to see more contributions from you.

Happy Contributing.

lasjorg reviewed (commented):
Why can't we just remove the whitespace? We have a helper for it as well.

```
const meta = [...document.querySelectorAll('meta')];
const target = meta?.find(
  (m) =>
    m?.getAttribute('name') === 'viewport' &&
    __helpers.removeWhiteSpace(m?.getAttribute('content')) ===
      'width=device-width,initial-scale=1.0'
);
assert.exists(target);
```

If we do need the regex, can we at least change the literal space match to a qualified whitespace match instead `\s*`

raisedadead commented:
@Saul-BT 

Can you update the PR in the next few days to remove white space instead as recommended in the comment above?

lasjorg commented:
I guess trimming the whitespace would technically allow for code with incorrect spaces to pass. I'm personally not too concerned about it but I can see it being a reason not to trim the whitespace. Not sure if anyone else has an opinion on it?

```
<meta name="viewport" content="width = device - width, initial - scale=1.0">
```

As said, if you do keep the regex please use `\s*` for the whitespace it just reads better and is more consistent.

---

As an aside. The way I first (mis)read the specs I was reading it as all the whitespace should be trimmed from the token. But I now see it is only leading and trailing whitespaces in the token ([split-on-commas](https://infra.spec.whatwg.org/#split-on-commas)).

I am a little surprised by how browsers handle it as no whitespace seems to be removed at all, including leading and trailing (in Chrome and Firefox). Code with incorrect spaces and without does not appear to produce the same result in the two browsers when tested using the browser's device mode. I'm not really sure how much I trust the device mode though.

The above code with incorrect spaces also passes validation which seems odd. So that is either not correct validation or the browsers are doing it wrong or I'm missing something.

Again, I'm not too concerned about it but I just wanted to mention it.

Saul-BT commented:
> As a new contributor, we encourage you to read our contributing guidelines.
> 
> We expect our contributors to be aware of the process specific to this project. Following the guidelines religiously earns you the respect of fellow maintainers and saves everyone time.
> 
> Some examples of this are:
> 
>     1. Do not edit off of GitHub–while you can, it's not a good idea.
> 
>     2. Make sure you follow the PR checklist and not just tick things off; otherwise, we won't take you seriously.
> 
>     3. Use the correct way to link PRs in the description of the PR by updating the `XXXXXX`. Do not just add issue numbers everywhere and anywhere you feel like.
> 
>     4. Use mentions in comments or review requests to a minimum or not. We understand you are excited about contributing, and our maintainers will get back to you as soon as they get a chance.
> 
> 
> We appreciate you taking the time to help us, and we hope to see more contributions from you.
> 
> Happy Contributing.

Hi, sorry for the late reply,

I understand what you are saying, and I apologize for it.

I will try to improve in the next contribution.

Thanks.

Saul-BT commented:
> I guess trimming the whitespace would technically allow for code with incorrect spaces to pass. I'm personally not too concerned about it but I can see it being a reason not to trim the whitespace. Not sure if anyone else has an opinion on it?
> 
> ```
> <meta name="viewport" content="width = device - width, initial - scale=1.0">
> ```
> 
> As said, if you do keep the regex please use `\s*` for the whitespace it just reads better and is more consistent.
> 
> As an aside. The way I first (mis)read the specs I was reading it as all the whitespace should be trimmed from the token. But I now see it is only leading and trailing whitespaces in the token ([split-on-commas](https://infra.spec.whatwg.org/#split-on-commas)).
> 
> I am a little surprised by how browsers handle it as no whitespace seems to be removed at all, including leading and trailing (in Chrome and Firefox). Code with incorrect spaces and without does not appear to produce the same result in the two browsers when tested using the browser's device mode. I'm not really sure how much I trust the device mode though.
> 
> The above code with incorrect spaces also passes validation which seems odd. So that is either not correct validation or the browsers are doing it wrong or I'm missing something.
> 
> Again, I'm not too concerned about it but I just wanted to mention it.

I'm also surprised at how the browsers handle it.

Anyway, I have performed several more tests both in Firefox and Chrome device mode, as well as on my own mobile.

I have observed the following:

I didn't look closely at the specification, but there they comment which characters are considered [whitespace](https://infra.spec.whatwg.org/#ascii-whitespace).

_I mistakenly assumed that line breaks would cause problems while parsing the value, but this makes it clear that they do not._

Therefore, I fully agree to replace ' ' with 's' ([reference](https://regex101.com/r/2z6dQs/1)).

Thank you all for the reviews 😃 .

Saul-BT committed:
Improve content value validation

Fix boundary bug using anchors `^` and `$` (to validate the full value)
Fix whitespace detection (https://infra.spec.whatwg.org/#ascii-whitespace)

Saul-BT commented:
In the last commit, I have added the following changes:

- The part '.0' of `initial-scale` is now optional ([reference](https://developer.mozilla.org/en-US/docs/Web/HTML/Viewport_meta_tag#screen_density)).

- The regex now works with the whitespace characters mentioned in the [specification](https://infra.spec.whatwg.org/#ascii-whitespace).

- I have added anchors to validate the whole content value, and not a substring.

lasjorg reviewed (approved):
LGTM
